File: packages/i18nify-js/src/modules/dateTime/getRelativeTime.ts
import { withErrorBoundary } from '../../common/errorBoundary';
import { getLocale } from '../.internal/utils';
import { DateInput, Locale } from './types';
import { convertToStandardDate } from './utils';

/**
 * Provides a relative time string (e.g., '3 hours ago', 'in 2 days').
 * This function calculates the difference between the given date and the base date,
 * then formats it in a locale-sensitive manner. It allows customization of the output
 * through Intl.RelativeTimeFormat options.
 *
 * @param date - The date to compare.
 * @param options - Config object.
 * @returns The relative time as a string.
 */
const getRelativeTime = (
  date: DateInput,
  options: {
    locale?: Locale;
    baseDate?: DateInput;
    intlOptions?: Intl.RelativeTimeFormatOptions;
  } = {},
): string => {
  const standardDate = convertToStandardDate(date);
  const standardBaseDate = convertToStandardDate(
    options.baseDate || new Date(),
  );

  const locale = getLocale(options);

  const diffInSeconds =
    (standardDate.getTime() - standardBaseDate.getTime()) / 1000;

  // Define time units in seconds
  const minute = 60;
  const hour = minute * 60;
  const day = hour * 24;
  const week = day * 7;
  const month = day * 30;
  const year = day * 365;

  let value: number;
  let unit: Intl.RelativeTimeFormatUnit;

  if (Math.abs(diffInSeconds) < minute) {
    value = diffInSeconds;
    unit = 'second';
  } else if (Math.abs(diffInSeconds) < hour) {
    value = diffInSeconds / minute;
    unit = 'minute';
  } else if (Math.abs(diffInSeconds) < day) {
    value = diffInSeconds / hour;
    unit = 'hour';
  } else if (Math.abs(diffInSeconds) < week) {
    value = diffInSeconds / day;
    unit = 'day';
  } else if (Math.abs(diffInSeconds) < month) {
    value = diffInSeconds / week;
    unit = 'week';
  } else if (Math.abs(diffInSeconds) < year) {
    value = diffInSeconds / month;
    unit = 'month';
  } else {
    value = diffInSeconds / year;
    unit = 'year';
  }

  let relativeTime;

  try {
    const rtf = new Intl.RelativeTimeFormat(locale, options.intlOptions);
    relativeTime = rtf.format(Math.round(value), unit);
  } catch (err) {
    if (err instanceof Error) {
      throw new Error(
        `An error occurred while creating the RelativeTimeFormat instance: ${err.message}. Please ensure the provided options are valid and try again.`,
      );
    } else {
      throw new Error(`An unknown error occurred. Error details: ${err}`);
    }
  }

  return relativeTime;
};

export default withErrorBoundary<typeof getRelativeTime>(getRelativeTime);

File: packages/i18nify-go/modules/date_time/get_relative_time.go
package date_time

import "fmt"

// GetRelativeTime calculates the relative time string (e.g., "2 hours ago" or "in 2 hours").
func GetRelativeTime(dateStr string, options DateTimeOptions) (string, error) {
	date, err := convertToStandardDate(dateStr)
	if err != nil {
		return "", err
	}
	baseDate, err := convertToStandardDate(options.BaseDate)
	if err != nil {
		return "", err
	}

	duration := date.Sub(baseDate)
	absDuration := duration
	prefix := "ago"

	// Handle future dates
	if duration > 0 {
		prefix = "in"
	} else {
		absDuration = -duration
	}

	// Format the duration with appropriate unit
	var result string
	switch {
	case absDuration < Minute:
		result = formatDuration(int(absDuration.Seconds()), "second", prefix)
	case absDuration < Hour:
		result = formatDuration(int(absDuration.Minutes()), "Minute", prefix)
	case absDuration < Day:
		result = formatDuration(int(absDuration.Hours()), "Hour", prefix)
	case absDuration < Week:
		result = formatDuration(int(absDuration.Hours()/24), "Day", prefix)
	case absDuration < Month:
		result = formatDuration(int(absDuration.Hours()/(24*7)), "Week", prefix)
	case absDuration < Year:
		result = formatDuration(int(absDuration.Hours()/(24*30)), "Month", prefix)
	default:
		result = formatDuration(int(absDuration.Hours()/(24*365)), "Year", prefix)
	}

	return result, nil
}

// formatDuration handles singular/plural forms and prefix placement
func formatDuration(value int, unit string, prefix string) string {
	if value == 0 {
		return fmt.Sprintf("0 %ss %s", unit, prefix)
	}

	// Handle singular/plural
	if value == 1 {
		if prefix == "ago" {
			return fmt.Sprintf("1 %s %s", unit, prefix)
		}
		return fmt.Sprintf("%s 1 %s", prefix, unit)
	}

	if prefix == "ago" {
		return fmt.Sprintf("%d %ss %s", value, unit, prefix)
	}
	return fmt.Sprintf("%s %d %ss", prefix, value, unit)
}
